Current calc correction

Invertedlight opened the pull request.

Invertedlight committed:
Update lifeblue.ino.ino

Added the Heltec.begin function call to the setup function to properly initialize the display.

Invertedlight committed:
Update BatteryManager.h

Corrected LIFE_SHORT_CIRCUITED bitmap to 0x20 per documentation.

Changed the 'current'  attribute from uint32_t to int32_t.  Needed so we can hold negative current values properly after 2's complement conversion during value read in  BatteryManager.cpp  function BatteryManager::convertBufferStringToValue.

Invertedlight committed:
Update BatteryManager.cpp

Added fix for negative values returned in the AMP field from the battery. Use 2's complement conversion when negative.
 BatteryManager::convertBufferStringToValue

Invertedlight commented:
Updated to do 2's complement on negative values returned by the battery.
Fixed the "#define LIFE_SHORT_CIRCUITED" to match the documentation bitmap setting.
Added Heltec.begin() call to initial setup() to activate the display properly.

coogle commented on BatteryManager.cpp:
```diff
@@ -184,8 +184,12 @@ uint32_t BatteryManager::convertBufferStringToValue(uint8_t len)
+      if (newNum < 0) {
+        newNum = ((newNum ^ 0xFFFFFFFF) +1 ) * -1;
+      }
+    return newNum; // END - Fix for negative AMP values -- JR
```
You can't do this here `newNum` is a `long` type but the `convertBufferStringToValue` method returns a `uint32_t` type.  To be consistent `long` should be `int32_t`, and the `convertBufferStringToValue` method needs to return `int32_t` instead of `uint32_t` as well.

coogle commented on lifeblue.ino.ino:
```diff
@@ -166,6 +166,9 @@ void setup() {
   }
+  
+  // Added Heltec display begin function – JR
+  Heltec.begin(true /*DisplayEnable Enable*/, false /*LoRa Enable*/, true /*Serial Enable*/);  
```
This certainly doesn't belong here. I'm not sure why you needed this, but at the very least it needs to exist in `DisplayManager::setup()`. See this line:

https://github.com/ThisSmartHouse/LiFeBlue-BLE/blob/59fedb0f9741d576fca883e4b7c2e487149555fd/DisplayManager.cpp#L91

I'm pretty sure the reason I added that line in the first place was to make the display work on LoRa devices (which I believe you are using). 

I looked at what exactly `Heltec.begin()` does in this case, and I'm wondering if that might offer some hints on how we can turn this on for this particular type of device?

https://github.com/HelTecAutomation/Heltec_ESP32/blob/c8ddaff755beeae15d22b216b67ece126b7c8fa5/src/heltec.cpp#L22

coogle reviewed (changes requested):
A few things need fixing..

Invertedlight commented:
Changing the return to uint32_t would require the other two values Voltage and FCC to be modified to uint32_t as well.

Would it be better to just move the 2’s complement code check up to the current assignment and leave the convertBufferStringToValue alone?



HelTec.begin()  - Makes sense to put it into the Display file. Keep things organized by functionality.
I was just stuck in the put crap in the initial setup Arduino thinking from all those examples I’ve poked at.

Will look at the display links you sent to check that out.


From: John Coggeshall <notifications@github.com>
Reply-To: ThisSmartHouse/LiFeBlue-BLE <reply@reply.github.com>
Date: Saturday, August 22, 2020 at 1:46 PM
To: ThisSmartHouse/LiFeBlue-BLE <LiFeBlue-BLE@noreply.github.com>
Cc: James Richardson <James.Richardson@invertedlight.com>, Author <author@noreply.github.com>
Subject: Re: [ThisSmartHouse/LiFeBlue-BLE] Current calc correction (#1)


@coogle requested changes on this pull request.

A few things need fixing..

________________________________

In BatteryManager.cpp<https://github.com/ThisSmartHouse/LiFeBlue-BLE/pull/1#discussion_r475135551>:

> @@ -184,8 +184,12 @@ uint32_t BatteryManager::convertBufferStringToValue(uint8_t len)



   switch(strlen(buf)) {



     case 4:



       return __builtin_bswap16(strtoul(buf, NULL, 16));



-    case 8:



-      return __builtin_bswap32(strtoul(buf, NULL, 16));



+    case 8: // Fix for negative AMP values -- JR



+      long newNum = __builtin_bswap32(strtoul(buf, NULL, 16));



+      if (newNum < 0) {



+        newNum = ((newNum ^ 0xFFFFFFFF) +1 ) * -1;



+      }



+    return newNum; // END - Fix for negative AMP values -- JR



You can't do this here newNum is a long type but the convertBufferStringToValue method returns a uint32_t type. To be consistent long should be int32_t, and the convertBufferStringToValue method needs to return int32_t instead of uint32_t as well.

________________________________

In lifeblue.ino.ino<https://github.com/ThisSmartHouse/LiFeBlue-BLE/pull/1#discussion_r475135921>:

> @@ -166,6 +166,9 @@ void setup() {



   for(int i = 0; (i < 50000) && !Serial; i++) {



     delay(1);



   }



+



+  // Added Heltec display begin function – JR



+  Heltec.begin(true /*DisplayEnable Enable*/, false /*LoRa Enable*/, true /*Serial Enable*/);



This certainly doesn't belong here. I'm not sure why you needed this, but at the very least it needs to exist in DisplayManager::setup(). See this line:

https://github.com/ThisSmartHouse/LiFeBlue-BLE/blob/59fedb0f9741d576fca883e4b7c2e487149555fd/DisplayManager.cpp#L91

I'm pretty sure the reason I added that line in the first place was to make the display work on LoRa devices (which I believe you are using).

I looked at what exactly Heltec.begin() does in this case, and I'm wondering if that might offer some hints on how we can turn this on for this particular type of device?

https://github.com/HelTecAutomation/Heltec_ESP32/blob/c8ddaff755beeae15d22b216b67ece126b7c8fa5/src/heltec.cpp#L22

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/ThisSmartHouse/LiFeBlue-BLE/pull/1#pullrequestreview-472960405>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AI6H2WE5YA3QTI75MBF2OZDSCAVDDANCNFSM4QIITH6A>.

coogle commented:
> Changing the return to uint32_t would require the other two values Voltage and FCC to be modified to uint32_t as well. Would it be better to just move the 2’s complement code check up to the current assignment and leave the convertBufferStringToValue alone?

No, it would require changes when we assign since those are never negative values, see:

https://github.com/ThisSmartHouse/LiFeBlue-BLE/blob/59fedb0f9741d576fca883e4b7c2e487149555fd/BatteryManager.cpp#L136-L144

Pretty sure we can get away with just casting voltage and such, but I'd confirm that of course:

```
currentBattery->voltage = (uint32_t) convertBufferStringToValue(8);
```

Invertedlight commented:
Makes sense.

What about the call when it is 4?  What does the  __builtin_bswap16 call return?
Can’t seem to find a good reference about that. My guess would be uint16_t which does not match.

switch(strlen(buf)) {
    case 4:
      return __builtin_bswap16(strtoul(buf, NULL, 16));
    case 8:
      return __builtin_bswap32(strtoul(buf, NULL, 16));
  }

From: John Coggeshall <notifications@github.com>
Reply-To: ThisSmartHouse/LiFeBlue-BLE <reply@reply.github.com>
Date: Saturday, August 22, 2020 at 2:00 PM
To: ThisSmartHouse/LiFeBlue-BLE <LiFeBlue-BLE@noreply.github.com>
Cc: James Richardson <James.Richardson@invertedlight.com>, Author <author@noreply.github.com>
Subject: Re: [ThisSmartHouse/LiFeBlue-BLE] Current calc correction (#1)


Changing the return to uint32_t would require the other two values Voltage and FCC to be modified to uint32_t as well. Would it be better to just move the 2’s complement code check up to the current assignment and leave the convertBufferStringToValue alone?

No, it would require changes when we assign since those are never negative values, see:

https://github.com/ThisSmartHouse/LiFeBlue-BLE/blob/59fedb0f9741d576fca883e4b7c2e487149555fd/BatteryManager.cpp#L136-L144

Pretty sure we can get away with just casting voltage and such, but I'd confirm that of course:

currentBattery->voltage = (uint32_t) convertBufferStringToValue(8);



—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/ThisSmartHouse/LiFeBlue-BLE/pull/1#issuecomment-678693907>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AI6H2WFD2XI5F4HYYU5C2RDSCAWV5ANCNFSM4QIITH6A>.

coogle commented:
> What about the call when it is 4? What does the __builtin_bswap16 call return? Can’t seem to find a good reference about that. My guess would be uint16_t which does not match.

Both of these `__builtin` functions are GCC-specific compiler functions that I would normally never ever use, but in this case the ESP32 *only* uses GCC for building stuff, so its probably okay.

The bottom of this page is where you can find them documented:

https://gcc.gnu.org/onlinedocs/gcc-4.8.2/gcc/Other-Builtins.html

It's okay that it returns uint16_t  because the extra bits are just going to be 0 and irrelevant:

```
    unsigned int foo = 16;
    printf("Foo: 0x%lx\n", (unsigned long)foo);
    printf("Foo: 0x%x\n", foo);
```

Both return `0x10`. Type casting can be weird sometimes, and compiler-dependent. In our case thankfully we only have one compiler to worry about so as long as it works I'm not concerned.

Invertedlight committed:
Updating Current signed value calculation

Cast to uint32_t since we changed the convertBufferStringToValue() to return int32_t for the signed Current value.

coogle commented:
I managed to get this running on my Heltec Lora 32 (v2), but the current readings I am getting from the batteries doesn't really look quite right. It is reporting a six-digit number for Amp draws, which obviously isn't right.

I need to dive into this patch a bit more and see what you're doing!

coogle commented:
Okay, so I actually got everything working and the only pressing problem with the PR at this point is the line in setup() where you initialize the display.

This line should be removed, and once it is I can merge this PR and we have made some real progress! It is possible that line is entirely unnecessary, assuming you are using the same ESP32 device as I am.

Make sure in Arduino for the Board you have "Heltec WiFi LoRa 32(v2)" selected. On my device this worked without needing any extra code for display initialization.

coogle commented:
Closing this PR because I took the meat of this, made it work locally on two separate ESP32 OLED devices and pushed the changes into master. 

Re-checkout `master` at this point and give it a try. It *should* work out of box as long as you select Heltec WiFi Lora32(V2) as the board (and set the partition scheme to minimal SPIFFS.

Invertedlight commented:
Thanks, I’ll give it a try and let you know.

From: John Coggeshall <notifications@github.com>
Reply-To: ThisSmartHouse/LiFeBlue-BLE <reply@reply.github.com>
Date: Thursday, September 10, 2020 at 1:32 PM
To: ThisSmartHouse/LiFeBlue-BLE <LiFeBlue-BLE@noreply.github.com>
Cc: James Richardson <James.Richardson@invertedlight.com>, Author <author@noreply.github.com>
Subject: Re: [ThisSmartHouse/LiFeBlue-BLE] Current calc correction (#1)


Closing this PR because I took the meat of this, made it work locally on two separate ESP32 OLED devices and pushed the changes into master.

Re-checkout master at this point and give it a try. It should work out of box as long as you select Heltec WiFi Lora32(V2) as the board (and set the partition scheme to minimal SPIFFS.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/ThisSmartHouse/LiFeBlue-BLE/pull/1#issuecomment-690710578>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AI6H2WHO2Q5MLYGFTMXUZZTSFEZV5ANCNFSM4QIITH6A>.

Invertedlight commented:
Question for you…

Looking at the circular buffer “currentBattery->buffer” object I can’t seem to get the system to like an index reference such as currentBattery->buffer[112]
The examples in the library show that    CircularBuffer<char, 50> buffer;  then    “buffer[2]; // ['c','d','e'] returns 'e'”

Yet I get an error trying to access the object’s buffer via an index reference:  currentBattery->buffer[112]

Any idea on how to make this happen?   All my attempts have failed with compile errors.
Creating a test file which defines a buffer like this:      (captured from the currentBattery->buffer as output)
char * buffer = "01D0D240D290D270D000000000000000000000000000000000000000000000000072D))))))))⸮91340000964B 000F8180300F2004100B40B008097901D0D240D290D270D0000000000000000000000000000000000000000000000000717))))))))⸮913400009D4B0000F8180300F20⸮93340000924B0000F8180300F2004100B20B00809790210D240D270D270D0000000000000000000000000000000000000000000000000715))))))))⸮93340000924B0000F8180300F2004100B40B00809690220D240D280D2B0D0000000000000000000000000000000000000000000000000731))))))))⸮99340000A54B0000F8180300F6004100B40B0080969⸮⸮⸮?⸮⸮⸮?⸮⸮⸮xV⸮⸮⸮⸮⸮?4⸮⸮L";

I can access the buffer[112] to find the good end value and calculate the checksum that matches the embedded checksum values.
My C++ appears to be quite rusty at this point.



Referencing currentBattery->buffer[112]   should give us the documentation end character/value to compare to the stated “0x41//END “ value.

We could exclude the section found by the 0x87 check of the end of buffer isn’t correct. Then scan further for the correct buffer string before adding in to the current battery batteryInfo_t structure.

That should eliminate the bad/whacky readings we see.



James


From: John Coggeshall <notifications@github.com>
Reply-To: ThisSmartHouse/LiFeBlue-BLE <reply@reply.github.com>
Date: Thursday, September 10, 2020 at 1:32 PM
To: ThisSmartHouse/LiFeBlue-BLE <LiFeBlue-BLE@noreply.github.com>
Cc: James Richardson <James.Richardson@invertedlight.com>, Author <author@noreply.github.com>
Subject: Re: [ThisSmartHouse/LiFeBlue-BLE] Current calc correction (#1)


Closing this PR because I took the meat of this, made it work locally on two separate ESP32 OLED devices and pushed the changes into master.

Re-checkout master at this point and give it a try. It should work out of box as long as you select Heltec WiFi Lora32(V2) as the board (and set the partition scheme to minimal SPIFFS.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/ThisSmartHouse/LiFeBlue-BLE/pull/1#issuecomment-690710578>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AI6H2WHO2Q5MLYGFTMXUZZTSFEZV5ANCNFSM4QIITH6A>.
